packages/core/core-v2/src/Parcel.js

```diff
@@ -0,0 +1,198 @@
+// @flow
+'use strict';
+import { AbortController } from 'abortcontroller-polyfill/dist/cjs-ponyfill';
+import Watcher from '@parcel/watcher';
+import PQueue from 'p-queue';
+import AssetGraph from './AssetGraph';
+import { Node } from './Graph';
+import type { Dependency, Asset, File } from './types';
+import TransformerRunner from './TransformerRunner';
+import ResolverRunner from './ResolverRunner';
+import BundlerRunner from './BundlerRunner';
+import PackagerRunner from './PackagerRunner';
+
+// TODO: use custom config if present
+const defaultConfig = require('@parcel/config-default');
+
+const abortError = new Error('Build aborted');
+
+type CliOpts = {
+  watch?: boolean
+}
+
+type ParcelOpts = {
+  entries: Array<string>,
+  cwd?: string,
+  cliOpts: CliOpts,
+}
+
+type Signal = {
+  aborted: boolean,
+  addEventListener?: Function,
+}
+
+type BuildOpts = {
+  signal: Signal,
+  shallow?: boolean,
+}
+
+export default class Parcel {
+  entries: Array<string>;
+  rootDir: string;
+  graph: AssetGraph;
+  watcher: Watcher;
+  updateQueue: PQueue;
+  mainQueue: PQueue;
+  transformerRunner: TransformerRunner;
+  resolverRunner: ResolverRunner;
+  bundlerRunner: BundlerRunner;
+  packagerRunner: PackagerRunner;
+
+  constructor({
+    entries,
+    cliOpts = {},
+  }: ParcelOpts) {
+    this.rootDir = process.cwd();
+
+    this.graph = new AssetGraph({ entries, rootDir: this.rootDir });
+    this.watcher = cliOpts.watch ? new Watcher() : null;
+    this.updateQueue = new PQueue({ autoStart: false });
+    this.mainQueue = new PQueue({ autoStart: false });
+
+    this.transformerRunner = new TransformerRunner({ parcelConfig: defaultConfig, cliOpts });
+    this.resolverRunner = new ResolverRunner();
+    this.bundlerRunner = new BundlerRunner({ parcelConfig: defaultConfig, cliOpts });
+    this.packagerRunner = new PackagerRunner({ parcelConfig: defaultConfig, cliOpts });
+  }
+
+  async run() {
+    let controller = new AbortController();
+    let signal = controller.signal;
+
+    let buildPromise = this.build({ signal });
+
+    if (this.watcher) {
+      this.watcher.on('change', event => {
+        // The build step starts by updating the graph with all of the changes in the update queue
+        if (!this.updateQueue.isPaused) {
+          // If a change comes in while the graph is being updated, just add it to the update queue
+          this.handleChange(event);
+        } else {
+          // Otherwise cancel any work that may be in progress, add the change to the update queue,
+          // and start a new build.
+          controller.abort();
+          this.mainQueue.pause();
+          this.mainQueue.clear();
+          this.handleChange(event);
+
+          controller = new AbortController();
+          signal = controller.signal;
+
+          this.build({ signal });
```

Comment:
I think we'll probably want some timeout here, so that if lots of files change at once (e.g. you switch git branches, or run npm install) we wait for things to cool down before trying to rebuild. See https://github.com/parcel-bundler/parcel/blob/master/src/Bundler.js#L769